File: src/test/kotlin/com/example/testDouble/mockk/StoreTest.kt
package com.example.testDouble.mockk

import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

@SpringBootTest
class StoreTest {

    @Autowired
    lateinit var store: Store

    @MockkBean
    lateinit var mockBackyard: Backyard

    @Test
    fun dummyTest() {
        // arrange
        every {
            mockBackyard.findItem(any())
        } throws RuntimeException()

        // act
        assertThrows<RuntimeException> {
            store.searchItem("")
        }
    }

    @Test
    fun spyTest() {
        // arrange
        val unusedReturnValue = ""
        every {
            mockBackyard.findItem(any())
        } returns unusedReturnValue

        // act
        val expectedItemName = "eternal youth"
        store.searchItem(expectedItemName)

        // assert
        verify {
            mockBackyard.findItem(expectedItemName)
        }
    }

    @Test
    fun stubTest() {
        // arrange
        val expectedItemLocation = "area 13"
        every {
            mockBackyard.findItem(any())
        } returns expectedItemLocation

        // act
        val unusedArgumentValue = ""
        val actualItemLocation = store.searchItem(unusedArgumentValue)

        // assert
        assertEquals(expectedItemLocation, actualItemLocation)
    }
}

File: react/src/test/Store.test.ts
import BackyardImpl from '@/main/Backyard.ts'
import StoreImpl from '@/main/Store.ts'
import { describe, expect, test, vi } from 'vitest'

vi.mock('@/main/Backyard')

describe('Store',()=>{
  test('dummyTest',()=>{
    // arrange
    const mockBackyard = new BackyardImpl()
    const mockFindItem = vi.spyOn(mockBackyard,'findItem')
    mockFindItem.mockImplementation(()=>{
      throw new Error()
    })
    const store = new StoreImpl(mockBackyard)

    // assert
    expect(()=>{
      // act
      const unusedArgumentValue = ''
      store.searchItem(unusedArgumentValue)
    }).toThrow(Error)
  })

  test('spyTest',()=>{
    // arrange
    const mockBackyard = new BackyardImpl()
    const mockFindItem = vi.spyOn(mockBackyard,'findItem')
    const store = new StoreImpl(mockBackyard)

    // act
    const expectedItemName = 'eternal youth'
    store.searchItem(expectedItemName)

    // assert
    expect(mockFindItem).toHaveBeenCalledWith(expectedItemName)
  })

  test('stubTest',()=>{
    // arrange
    const mockBackyard = new BackyardImpl()
    const mockFindItem = vi.spyOn(mockBackyard,'findItem')
    const expectedItemLocation = 'area 13'
    mockFindItem.mockImplementation(()=>{
      return expectedItemLocation
    })
    const store = new StoreImpl(mockBackyard)

    // act
    const unusedArgumentValue = ''
    const actualItemLocation = store.searchItem(unusedArgumentValue)

    // assert
    expect(actualItemLocation).toEqual(expectedItemLocation)
  })
})